### DISCUSSION TITLE
more helpful Unsat report

### DISCUSSION THREAD

**Discussion by jmarca** [google/or-tools#3001] [CP-SAT questions]
**What language and solver does this apply to?**
All, CPSat

**Describe the problem you are trying to solve.**

Sometimes my code is buggy, and the model is unsatisfiable.  It is hard to figure out which constraint is broken.

The output looks like:

> Starting presolve at 0.00s
> Unsat after presolving constraint #2999 (warning, dump might be inconsistent): linear { vars: 421 vars: 675 vars: 812 vars: 1066 coeffs: 1 coeffs: 1 coeffs: 1 coeffs: 1 domain: -9223372036854775808 domain: -1 }

Usually I *think* I know where the problem is, but with thousands of constraints it is hard to be sure.

**Describe the solution you'd like**

I am pretty careful to always give the string names to constraints and the like.  Would be nice to see that.

Something like:

> Starting presolve at 0.00s
> Unsat after presolving constraint #2999 "Total souls assigned for demon Bob" (warning, dump might be inconsistent): linear { vars: 421 vars: 675 vars: 812 vars: 1066 coeffs: 1 coeffs: 1 coeffs: 1 coeffs: 1 domain: -9223372036854775808 domain: -1 }

**Describe alternatives you've considered**

Liberal print statements prior to calling solve.  Careful analysis of the model.

**Additional context**

**Comment by gregy4** [comment]
Maybe only programatic access to what is written is needed. Sometimes it can be helpful to not just display a message but to do some action.

jmarca: do you know assumptions ?  in short, when you rewrite all constraints to be conditional via channeling constraints, you can get minimal set of infeasible constraints. boolean variables used in enforcements should be marked as an assumption. alternative model with assumptions can be helpful to catch infeasibility in input data.

**Comment by jmarca** [comment]
I'm not sure that is relevant to what I'm asking.

When I create a variable, etc, I do:

```
var roomsCleaned = model.NewIntVar(minRooms, maxRooms, $"atndnt {attendant.Username}: assigned rooms");
```

All I want is that string at the end of the NewIntVar, etc to show up in the diagnostic output.

So when the constraint fails, the error says something that I can trace back in my code.  Constraint # 2345 is not helpful.

**Comment by jmarca** [comment]
To state another way, I have a large solver project that is growing organically with additional requirements and conditions, etc.  If I just graft on a new feature, no big deal, it usually works or fails fast and I figure out what I've done wrong.  But as with any big project, occasionally the benefits of refactoring outweigh the pains, so I refactor.  This is when the mysterious error message drive me crazy, as I just touched lots of code and have no idea which part of the input conditions and constraints I've screwed up. 

In this case I fixed it and moved on, and I don't even remember what the ultimate bug was.  I'm not arguing that it is some solver issue...it is absolutely a mistake on my part.  

I'm not going to get anywhere by reformulating as channeling constraints etc., because exactly that sort of reformulating exercise is what got me into trouble in the first place!  I just want to find those mistakes faster when I'm refactoring/reformulating.

**Comment by lperron** [comment]
I agree we could improved model debugging.
It is progressing when we need to debug our models :-)

**Comment by gregy4** [comment]
> I'm not sure that is relevant to what I'm asking.
> 
> When I create a variable, etc, I do:
> 
> ```
> var roomsCleaned = model.NewIntVar(minRooms, maxRooms, $"atndnt {attendant.Username}: assigned rooms");
> ```
> 
> All I want is that string at the end of the NewIntVar, etc to show up in the diagnostic output.
> 
> So when the constraint fails, the error says something that I can trace back in my code. Constraint # 2345 is not helpful.

You can trace it, since you can get the constraint by the index. Same is true for variables.
In case a method to programmatically get index of failed constraint and its variables will be available, it should be easy to produce string output that you want.

**Comment by jmarca** [comment]
@gregy4 perhaps you are missing the issue tag of "Feature Request"?

**Comment by fdid-git** [comment]
Note that if the constraint is named, it should appear in the log message already with name:"xxx".
I suspect the failing constraint is one we create internally, it is why there is no name.

But yes, we are aware that the current messages are not the best :)
